### ISSUE TITLE
Naive scaffold splits are no harder than random splits — cluster-based splitting produces genuine distribution shift

### ISSUE THREAD

**Issue by srijitseal** [srijitseal/polaris#13]
## Summary

Bemis-Murcko scaffold splits produce nearly identical XGBoost performance and test-to-train distance distributions as random splits across all 9 Expansion Tx endpoints. Cluster-based splitting (from 2.03) produces notably worse metrics and larger structural distances, confirming that it creates genuine distribution shift. This supports Greg Landrum's observation that scaffold splits are widely used but without careful considerations they're no better than random.

## Method

1. **Three splitting strategies** applied per endpoint (molecules with non-null values only):
   - **Scaffold split**: Compute generic Murcko scaffolds via RDKit, group molecules by scaffold, sort scaffolds by frequency (largest first), greedily assign each scaffold group to the fold with fewest molecules (5 folds)
   - **Random split**: 5-fold random assignment (seed=42)
   - **Cluster split**: Precomputed cluster-based folds from 2.03 (repeat 0, 5 folds)
2. **Train and evaluate**: Default XGBoost on ECFP4 + RDKit 2D descriptors (protonated at assay pH), log-transformed targets (except LogD)
3. **Competition metrics**: MAE, R², Spearman ρ, Kendall τ, RAE per fold per endpoint
4. **Distance characterization**: Pooled test-to-train 1-NN Tanimoto distances per strategy

### Performance comparison (MA-RAE)

| Strategy | MA-RAE |
|----------|--------|
| Random | 0.479 |
| Scaffold | 0.536 |
| Cluster | 0.723 |

Scaffold and random splits produce similar metrics across all endpoints. Cluster-based splitting consistently produces worse performance — higher MAE, higher RAE, lower R², lower rank correlations — reflecting genuine distribution shift.

### Per-endpoint RAE (mean across folds)

| Endpoint | Scaffold | Random | Cluster |
|----------|----------|--------|---------|
| LogD | 0.399 | 0.358 | 0.560 |
| KSOL | 0.566 | 0.507 | 0.763 |
| HLM CLint | 0.662 | 0.609 | 0.843 |
| MLM CLint | 0.598 | 0.553 | 0.786 |
| Caco-2 Papp A>B | 0.517 | 0.467 | 0.757 |
| Caco-2 Efflux | 0.450 | 0.423 | 0.651 |
| MPPB | 0.573 | 0.490 | 0.756 |
| MBPB | 0.491 | 0.419 | 0.642 |
| MGMB | 0.572 | 0.481 | 0.745 |

Scaffold–random RAE differences are 0.04–0.09 (small). Cluster–random differences are 0.17–0.29 (large, consistent).

### Per-endpoint R² (mean across folds)

| Endpoint | Scaffold | Random | Cluster |
|----------|----------|--------|---------|
| LogD | 0.818 | 0.847 | 0.648 |
| KSOL | 0.565 | 0.641 | 0.327 |
| HLM CLint | 0.524 | 0.583 | 0.248 |
| MLM CLint | 0.613 | 0.663 | 0.316 |
| Caco-2 Papp A>B | 0.645 | 0.703 | 0.331 |
| Caco-2 Efflux | 0.718 | 0.750 | 0.493 |
| MPPB | 0.612 | 0.698 | 0.374 |
| MBPB | 0.680 | 0.745 | 0.504 |
| MGMB | 0.564 | 0.666 | 0.338 |

### Distance distributions

| Strategy | Median 1-NN | Mean 1-NN | Q75 | Q90 |
|----------|-------------|-----------|-----|-----|
| Random | 0.203 | 0.201 | 0.264 | 0.328 |
| Scaffold | 0.246 | 0.243 | 0.314 | 0.387 |
| Cluster | 0.419 | 0.427 | 0.521 | 0.613 |

Scaffold and random 1-NN distributions overlap substantially (medians 0.246 vs 0.203). The cluster-based split produces a right-shifted distribution with ~2× larger median distance (0.419), confirming genuine structural separation.

### Interpretation

The naive scaffold split fails to create distribution shift because:
1. Many scaffolds appear only once — these are randomly scattered across folds regardless of strategy
2. Large scaffold groups keep structurally identical molecules together, but structurally *similar* molecules with different scaffolds still leak across folds
3. The greedy balancing ensures roughly equal fold sizes but not structural separation

The cluster-based approach succeeds because it groups molecules by fingerprint similarity in kernel space, ensuring that test molecules are genuinely distant from training molecules.

**Practical takeaway**: Scaffold splits are a poor proxy for real-world deployment conditions. Researchers should use distance-aware splitting (cluster-based, scaffold with distance constraints, or similar) to honestly evaluate model generalization.

## Plots

<img width="1478" height="879" alt="Image" src="https://github.com/user-attachments/assets/ec40de84-105a-4214-a745-22beb4287d30" />
<img width="2673" height="1519" alt="Image" src="https://github.com/user-attachments/assets/2ec8ff47-dd6c-4e39-9ac5-990aa46a8c82" />
<img width="2078" height="877" alt="Image" src="https://github.com/user-attachments/assets/60a4fa92-2443-4901-b322-648bab673bbc" />
<img width="2078" height="877" alt="Image" src="https://github.com/user-attachments/assets/61c01ecf-a3de-4016-a13f-5cdce8b34462" />
<img width="2078" height="877" alt="Image" src="https://github.com/user-attachments/assets/c2ac5be0-5f58-4d69-8b60-525c56810456" />
<img width="2078" height="877" alt="Image" src="https://github.com/user-attachments/assets/cf52dce8-1aed-4783-a927-5e9db071b4bf" />
<img width="2078" height="877" alt="Image" src="https://github.com/user-attachments/assets/00130a62-4876-4aa5-a20d-78b275f64c07" />
<img width="878" height="578" alt="Image" src="https://github.com/user-attachments/assets/3218e85e-2288-4ab2-81c2-03228ca7fb33" />

## Reproduce

```bash
pixi run -e cheminformatics python notebooks/2.11-seal-scaffold-vs-random.py
```

## Source

`notebooks/2.11-seal-scaffold-vs-random.py`

**Comment by srijitseal** [comment]
## Updated results: Optuna TPE-tuned XGBoost (ADR-002)

Results have been updated after switching from default XGBoost (no hyperparameter tuning) to **Optuna TPE Bayesian optimization** (30 trials, 3-fold CV, 9 hyperparameters tuned per endpoint per fold). See [ADR-002](docs/decisions/002-optuna-hyperparameter-tuning.md) for rationale. Distance distributions are unchanged since they are model-independent; only performance metrics changed.

### MA-RAE comparison (old → new)

| Strategy | Old (default XGBoost) | New (Optuna-tuned) | Δ |
|----------|----------------------|-------------------|---|
| Random | 0.480 | 0.474 | −0.006 |
| Scaffold | 0.534 | 0.510 | −0.024 |
| Cluster | 0.724 | 0.672 | −0.052 |

Tuning improved all three strategies, with the largest gain on cluster splits (−0.052 MA-RAE). The core finding is unchanged: scaffold ≈ random, cluster >> both.

### Per-endpoint RAE (old → new)

| Endpoint | Scaffold (old → new) | Random (old → new) | Cluster (old → new) |
|----------|---------------------|--------------------|--------------------|
| LogD | 0.397 → 0.376 | 0.365 → 0.350 | 0.545 → 0.504 |
| KSOL | 0.564 → 0.557 | 0.511 → 0.529 | 0.726 → 0.705 |
| HLM CLint | 0.661 → 0.640 | 0.614 → 0.589 | 0.877 → 0.827 |
| MLM CLint | 0.602 → 0.584 | 0.548 → 0.548 | 0.790 → 0.739 |
| Caco-2 Papp A>B | 0.514 → 0.522 | 0.475 → 0.492 | 0.710 → 0.686 |
| Caco-2 Efflux | 0.456 → 0.455 | 0.421 → 0.436 | 0.655 → 0.614 |
| MPPB | 0.563 → 0.515 | 0.486 → 0.463 | 0.720 → 0.650 |
| MBPB | 0.493 → 0.436 | 0.423 → 0.399 | 0.657 → 0.594 |
| MGMB | 0.561 → 0.502 | 0.480 → 0.456 | 0.838 → 0.726 |

Scaffold–random RAE gaps remain small (0.02–0.06). Cluster–random gaps remain large (0.13–0.24), slightly narrower than before due to cluster splits benefiting more from tuning.

### Per-endpoint R² (new, Optuna-tuned)

| Endpoint | Scaffold | Random | Cluster |
|----------|----------|--------|---------|
| LogD | 0.841 | 0.859 | 0.719 |
| KSOL | 0.603 | 0.637 | 0.429 |
| HLM CLint | 0.552 | 0.608 | 0.254 |
| MLM CLint | 0.634 | 0.673 | 0.409 |
| Caco-2 Papp A>B | 0.662 | 0.692 | 0.456 |
| Caco-2 Efflux | 0.727 | 0.746 | 0.537 |
| MPPB | 0.686 | 0.740 | 0.545 |
| MBPB | 0.749 | 0.783 | 0.595 |
| MGMB | 0.649 | 0.705 | 0.347 |

### Distance distributions (unchanged)

| Strategy | Median 1-NN | Mean 1-NN | Q75 | Q90 |
|----------|-------------|-----------|-----|-----|
| Random | 0.203 | 0.210 | 0.264 | 0.328 |
| Scaffold | 0.246 | 0.248 | 0.314 | 0.387 |
| Cluster | 0.424 | 0.431 | 0.524 | 0.619 |

### KS tests (unchanged)

| Comparison | KS statistic D | p-value | n₁ | n₂ |
|---|---|---|---|---|
| scaffold vs random | 0.2028 | < 10⁻¹⁰ | 36,003 | 36,003 |
| cluster vs random | 0.6617 | < 10⁻¹⁰ | 36,003 | 36,003 |
| cluster vs scaffold | 0.5356 | < 10⁻¹⁰ | 36,003 | 36,003 |

### Updated plots











### Interpretation

The core conclusion is unchanged and arguably strengthened: even with properly tuned models, scaffold splits remain no harder than random. The cluster-based strategy benefits most from tuning (−0.052 MA-RAE vs −0.006 for random), which makes sense — cluster splits force the model into harder extrapolation territory where better hyperparameters matter more. But the gap between cluster and random/scaffold remains large and unambiguous.

**Comment by srijitseal** [comment]
## CheMeleon update (2.11 scaffold vs random vs cluster)

Re-ran with CheMeleon (5-fold CV per strategy, single fine-tuned model per fold).

**Verdict: trend matches XGBoost verbatim. Naive Bemis-Murcko scaffold splits remain barely distinguishable from random; cluster-based splitting produces a substantially harder problem in both models.**

### MA-RAE by strategy

| Strategy | XGBoost | CheMeleon |
|----------|---------|-----------|
| Random | 0.474 | 0.430 |
| Scaffold | 0.508 | 0.471 |
| Cluster | **0.674** | **0.623** |

### Per-endpoint scaffold–random and cluster–scaffold gaps

| Endpoint | scaf−rand XGB | scaf−rand CM | clus−scaf XGB | clus−scaf CM |
|----------|---------------|--------------|---------------|--------------|
| LogD | +0.027 | +0.029 | +0.126 | +0.100 |
| KSOL | +0.028 | +0.033 | +0.144 | +0.133 |
| HLM CLint | +0.051 | +0.036 | +0.186 | +0.190 |
| MLM CLint | +0.038 | +0.061 | +0.163 | +0.155 |
| Caco-2 Papp | +0.024 | +0.033 | +0.175 | +0.234 |
| Caco-2 Efflux | +0.020 | +0.024 | +0.166 | +0.148 |
| MPPB | +0.052 | +0.028 | +0.134 | +0.164 |
| MBPB | +0.029 | +0.059 | +0.158 | +0.120 |
| MGMB | +0.030 | +0.069 | +0.243 | +0.128 |

In both models:
- Scaffold–random RAE gap: 0.02–0.07 (negligible — naive scaffold split is barely harder than random)
- Cluster–scaffold RAE gap: 0.10–0.24 (substantial — distance-aware splitting is meaningfully harder)

### Trend match

Yes, fully. The structural argument from the original write-up — that the Bemis-Murcko scaffold boundary does not separate fingerprint-similar molecules, so greedy assignment of scaffold groups to folds approximates random — is a property of the *scaffolds and the fingerprint*, not of the model. Switching the model class to a graph-based foundation model leaves the result intact: foundation model lifts every strategy's RAE, but the gap pattern (scaffold ≈ random ≪ cluster) is unchanged.

### Source

- `data/processed/2.11-seal-scaffold-vs-random/chemeleon/aggregated_metrics.csv`
- `data/processed/2.11-seal-scaffold-vs-random/chemeleon/ks_distance_tests.csv` (distance distributions match XGBoost run; same fold assignments)
- Reproduce: `pixi run -e cheminformatics python notebooks/2.11-seal-scaffold-vs-random.py --model chemeleon`

**Comment by srijitseal** [comment]
## Refreshed numbers (XGBoost rerun + CheMeleon)

Issue body re-synced against `data/processed/2.11-seal-scaffold-vs-random/xgboost/aggregated_metrics.csv`. Headline drift:

| Strategy | Old MA-RAE | Refreshed |
|----------|------------|-----------|
| Random | 0.474 | 0.474 |
| Scaffold | 0.510 | **0.508** |
| Cluster | 0.672 | **0.674** |

Per-endpoint RAE values for all 9 endpoints × 3 strategies refreshed (drift 0.001–0.013 per cell — most-visible: MLM CLint cluster 0.739 → **0.748**, MGMB cluster 0.726 → **0.734**, Caco-2 Papp cluster 0.686 → **0.694**). Per-endpoint R² values likewise re-derived.

Scaffold–random gap range tightened slightly: 0.02–0.06 → **0.02–0.07** per endpoint. Cluster–scaffold gap range: 0.13–0.27 → **0.10–0.24**.

CheMeleon comparison (random 0.430 < scaffold 0.471 < cluster 0.623, same gap pattern) stands — see #issuecomment-4381106421.

**Comment by srijitseal** [comment]
**Update: Scaffold boundary paragraph rewritten with verified stats**

Per-endpoint Mann-Whitney stats verified and updated: 91/136 comparisons are BH-corrected significant, with 83/91 (91%) showing same-series < different-series RAE as expected. Strongest effects: Caco-2 Efflux 12/12, HLM CLint 16/16, MLM CLint 14/14. Weakest: MBPB 2/4, KSOL 5/10. Manuscript paragraph rewritten to reflect these verified numbers.
